In [ ]:
import math
import numpy as np
import time
import matplotlib.pyplot as plt
import pandas as pd
import pickle

<b> Corrections (PS6) </br>
<p>Q3: -2 for too long variable names and lines exceed 80</p>

**Q3.**

In [ ]:
#a. 
data12 = pd.read_sas("/DEMO_G.XPT")
data34 = pd.read_sas("/DEMO_H.XPT")
data56 = pd.read_sas("/DEMO_I.XPT")
data78 = pd.read_sas("/DEMO_J.XPT")

#Add an id to each cohort 
data12["Cohort_ID"] = [1 for i in range(data12.shape[0])]
data34["Cohort_ID"] = [2 for i in range(data34.shape[0])]
data56["Cohort_ID"] = [3 for i in range(data56.shape[0])]
data78["Cohort_ID"] = [4 for i in range(data78.shape[0])]

data1 = data12[["SEQN", "RIDAGEYR", "RIDRETH3", "DMDEDUC2", "DMDMARTL", 
                   "RIDSTATR", "SDMVPSU", "SDMVSTRA", "WTMEC2YR", "WTINT2YR", "Cohort_ID"]] 
data2 = data34[["SEQN", "RIDAGEYR", "RIDRETH3", "DMDEDUC2", "DMDMARTL", 
                   "RIDSTATR", "SDMVPSU", "SDMVSTRA", "WTMEC2YR", "WTINT2YR", "Cohort_ID"]] 
data3 = data56[["SEQN", "RIDAGEYR", "RIDRETH3", "DMDEDUC2", "DMDMARTL", 
                   "RIDSTATR", "SDMVPSU", "SDMVSTRA", "WTMEC2YR", "WTINT2YR", "Cohort_ID"]] 
data4 = data78[["SEQN", "RIDAGEYR", "RIDRETH3", "DMDEDUC2", "DMDMARTL", 
                   "RIDSTATR", "SDMVPSU", "SDMVSTRA", "WTMEC2YR", "WTINT2YR", "Cohort_ID"]]

#append 
frame = [data1, data2, data3, data4]
data_clean = pd.concat(frame)

#rename the columns 
data_clean = data_clean.rename(columns = {"SEQN": "respondent sequence number", "RIDAGEYR": 
                                                "age in years at screening", 
                                                "RIDRETH3": "race/hispanic origin w/ nh asian", 
                                                "DMDEDUC2": "education level - adults 20+", 
                                                "DMDMARTL": "marital status", 
                                                "RIDSTATR": "interview/examination status", 
                                                "SDMVPSU": "masked variance pseudo-PSU", 
                                                "SDMVSTRA": "masked variance pseudo-stratum", 
                                                "WTMEC2YR": "full sample 2 year mec exam weight", 
                                                "WTINT2YR": "full sample 2 year interview weight"})

#replace NaN with -1
data_clean = data_clean.fillna(-1)

#convert the data structure type in each column
convert_data = {}
for i in data_clean.columns:
    if i != "WTMEC2YR" and i != "WTINT2YR":
        convert_data[i] = int

data_clean = data_clean.round(2) #to round WTMEC2YR and WTINT2YR to two decimal digits 
data_clean = data_clean.astype(convert_data)

#serialize the data with pickle
serial_data = pickle.dumps(data_clean)
pickle.loads(serial_data)

respondent sequence number  ...  Cohort_ID
0                          62161  ...          1
1                          62162  ...          1
2                          62163  ...          1
3                          62164  ...          1
4                          62165  ...          1
...                          ...  ...        ...
9249                      102952  ...          4
9250                      102953  ...          4
9251                      102954  ...          4
9252                      102955  ...          4
9253                      102956  ...          4

[39156 rows x 11 columns]

In [ ]:
#b. 

data12_oral = pd.read_sas("/OHXDEN_G.XPT")
data34_oral = pd.read_sas("/OHXDEN_H.XPT")
data56_oral = pd.read_sas("/OHXDEN_I.XPT")
data78_oral = pd.read_sas("/OHXDEN_J.XPT")

#Add an id to each cohort 
data12_oral["Cohort_ID"] = [1 for i in range(data12_oral.shape[0])]
data34_oral["Cohort_ID"] = [2 for i in range(data34_oral.shape[0])]
data56_oral["Cohort_ID"] = [3 for i in range(data56_oral.shape[0])]
data78_oral["Cohort_ID"] = [4 for i in range(data78_oral.shape[0])]

#Obtain the column names
column_names = ["SEQN", "OHDDESTS"]
for i in range(1, 33):
    if i < 10:
        temp = "OHX0" + str(i) + "TC"
        column_names.append(temp)
    else:
        temp = "OHX" + str(i) + "TC"
        column_names.append(temp)

for i in range(2, 32):
    if i != 17 and i != 16:
        if i < 10:
            temp = "OHX0" + str(i) + "CTC"
            column_names.append(temp)
        else:
            temp = "OHX" + str(i) + "CTC"
            column_names.append(temp)

#extract columns from data        
data1_oral = data12_oral[column_names] 
data2_oral = data34_oral[column_names] 
data3_oral = data56_oral[column_names] 
data4_oral = data78_oral[column_names] 

#append 
frame_oral = [data1_oral, data2_oral, data3_oral, data4_oral]
data_oral = pd.concat(frame_oral)

#rename the column names
data_oral = data_oral.rename(columns = {"OHX01TC": "Tooth Count: Upper right 3rd molar (3M)",
                                                         "OHX02TC": "Tooth Count: Upper right 2nd molar (2M)",
                                                         "OHX03TC": "Tooth Count: Upper right 1st molar (1M)",
                                                         "OHX04TC": "Tooth Count: Upper right 2nd bicuspid/2nd primary molar (2B)",
                                                         "OHX05TC": "Tooth Count: Upper right 1st bicuspid/1st primary molar (1B)",
                                                         "OHX06TC": "Tooth Count: Upper right cuspid (C)",
                                                         "OHX07TC": "Tooth Count: Upper right lateral incisor (LI)",
                                                         "OHX08TC": "Tooth Count: Upper right central incisor (CI)",
                                                         "OHX09TC": "Tooth Count: Upper left central incisor (CI)",
                                                         "OHX10TC": "Tooth Count: Upper left lateral incisor (LI)",
                                                         "OHX11TC": "Tooth Count: Upper left cuspid (C)",
                                                         "OHX12TC": "Tooth Count: Upper left 1st bicuspid/1st primary molar (1B)",
                                                         "OHX13TC": "Tooth Count: Upper left 2nd bicuspid/2nd primary molar (2B)",
                                                         "OHX14TC": "Tooth Count: Upper left 1st molar (1M)",
                                                         "OHX15TC": "Tooth Count: Upper left 2nd molar (2M)",
                                                         "OHX16TC": "Tooth Count: Upper left 3rd molar (3M)",
                                                         "OHX17TC": "Tooth Count: Lower left 3rd molar (3M)",
                                                         "OHX18TC": "Tooth Count: Lower left 2nd molar (2M)",
                                                         "OHX19TC": "Tooth Count: Lower left 1st molar (1M)",
                                                         "OHX20TC": "Tooth Count: Lower left 2nd bicuspid/2nd primary molar (2B)",
                                                         "OHX21TC": "Tooth Count: Lower left 1st bicuspid/1st primary molar (1B)",
                                                         "OHX22TC": "Tooth Count: Lower left cuspid (C)",
                                                         "OHX23TC": "Tooth Count: Lower left lateral incisor (LI)",
                                                         "OHX24TC": "Tooth Count: Lower left central incisor (CI)",
                                                         "OHX25TC": "Tooth Count: Lower right central incisor (CI)",
                                                         "OHX26TC": "Tooth Count: Lower right lateral incisor (LI)",
                                                         "OHX27TC": "Tooth Count: Lower right cuspid (C)",
                                                         "OHX28TC": "Tooth Count: Lower right 1st bicuspid/1st primary molar (1B)",
                                                         "OHX29TC": "Tooth Count: Lower right 2nd bicuspid/2nd primary molar (2B)",
                                                         "OHX30TC": "Tooth Count: Lower right 1st molar (1M)",
                                                         "OHX31TC": "Tooth Count: Lower right 2nd molar (2M)",
                                                         "OHX32TC": "Tooth Count: Lower right 3rd molar (3M)"})
data_oral = data_oral.rename(columns = {"OHX02CTC": "Coronal Caries: Upper right 2nd molar (2M) tooth code",
                                                         "OHX03CTC": "Coronal Caries: Upper right 1st molar (1M) tooth code",
                                                         "OHX04CTC": "Coronal Caries: Upper right 2nd bicuspid/2nd primary molar (2B) tooth code",
                                                         "OHX05CTC": "Coronal Caries: Upper right 1st bicuspid/1st primary molar (1B) tooth code",
                                                         "OHX06CTC": "Coronal Caries: Upper right cuspid (C) tooth code",
                                                         "OHX07CTC": "Coronal Caries: Upper right lateral incisor (LI) tooth code",
                                                         "OHX08CTC": "Coronal Caries: Upper right central incisor (CI) tooth code",
                                                         "OHX09CTC": "Coronal Caries: Upper left central incisor (CI) tooth code",
                                                         "OHX10CTC": "Coronal Caries: Upper left lateral incisor (LI) tooth code",
                                                         "OHX11CTC": "Coronal Caries: Upper left cuspid (C) tooth code",
                                                         "OHX12CTC": "Coronal Caries: Upper left 1st bicuspid/1st primary molar (1B) tooth code",
                                                         "OHX13CTC": "Coronal Caries: Upper left 2nd bicuspid/2nd primary molar (2B) tooth code",
                                                         "OHX14CTC": "Coronal Caries: Upper left 1st molar (1M) tooth code",
                                                         "OHX15CTC": "Coronal Caries: Upper left 2nd molar (2M) tooth code",
                                                         "OHX18CTC": "Coronal Caries: Lower left 2nd molar (2M) tooth code",
                                                         "OHX19CTC": "Coronal Caries: Lower left 1st molar (1M) tooth code",
                                                         "OHX20CTC": "Coronal Caries: Lower left 2nd bicuspid/2nd primary molar (2B) tooth code",
                                                         "OHX21CTC": "Coronal Caries: Lower left 1st bicuspid/1st primary molar (1B) tooth code",
                                                         "OHX22CTC": "Coronal Caries: Lower left cuspid (C) tooth code",
                                                         "OHX23CTC": "Coronal Caries: Lower left lateral incisor (LI) tooth code",
                                                         "OHX24CTC": "Coronal Caries: Lower left central incisor (CI) tooth code",
                                                         "OHX25CTC": "Coronal Caries: Lower right central incisor (CI) tooth code",
                                                         "OHX26CTC": "Coronal Caries: Lower right lateral incisor (LI) tooth code",
                                                         "OHX27CTC": "Coronal Caries: Lower right cuspid (C) tooth code",
                                                         "OHX28CTC": "Coronal Caries: Lower right 1st bicuspid/1st primary molar (1B) tooth code",
                                                         "OHX29CTC": "Coronal Caries: Lower right 2nd bicuspid/2nd primary molar (2B) tooth code",
                                                         "OHX30CTC": "Coronal Caries: Lower right 1st molar (1M) tooth code",
                                                         "OHX31CTC": "Coronal Caries: Lower right 2nd molar (2M) tooth code"})
data_oral = data_oral.rename(columns = {"SEQN": "respondent sequence number", 
                                                          "OHDDESTS": "Dentition Status Code"})      
for i in data_oral.columns:
    data_oral = data_oral.rename(columns = {i: i.lower()})  #convert the column names to lowercase                                

#replace NaN with -1
data_oral = data_oral.fillna(-1)

#convert the data structure type in each column
converted_data = []
convert = {}
for i in data_oral.columns:
    series = data_oral[i]
    str_type = pd.api.types.is_string_dtype(series)
    num_type = pd.api.types.is_numeric_dtype(series)
    
    if str_type == True:   
        temp = data_oral.get(i).str.decode("utf-8").replace({"A": "Primary tooth with a restored surface condition",
                                                                      "D": "Sound primary tooth",
                                                                      "E": "Missing due to dental disease",
                                                                      "F": "Permanent tooth with a restored surface condition",
                                                                      "J": "Permanent root tip is present but no restorative replacement is present",
                                                                      "K": "Primary tooth with a dental carious surface condition",
                                                                      "M": "Missing due to other causes	",
                                                                      "P": "Missing due to dental disease but replaced by a removable restoration",
                                                                      "Q": "Missing due to other causes but replaced by a removable restoration",
                                                                      "R": "Missing due to dental disease but replaced by a fixed restoration",
                                                                      "S": "Sound permanent tooth",
                                                                      "T": "Permanent root tip is present but a restorative replacement is present",
                                                                      "U": "Unerupted",
                                                                      "X": "Missing due to other causes but replaced by a fixed restoration",
                                                                      "Y": "Tooth present, condition cannot be assessed",
                                                                      "Z": "Permanent tooth with a dental carious surface condition",
                                                                      " ": "Missing"})
        converted_data.append(temp)
    if num_type == True:
        convert[i] = int
        temp = data_oral.get(i)
        converted_data.append(temp)

converted_frame = pd.DataFrame()
k = 0
for i in data_oral.columns:
    converted_frame[i] = converted_data[k]
    k = k + 1
converted_frame = converted_frame.astype(convert)

#serialize the data with pickle
serial_data = pickle.dumps(converted_frame)
pickle.loads(serial_data)


respondent sequence number  ...  coronal caries: lower right 2nd molar (2m) tooth code
0                          62161  ...                              Sound permanent tooth    
1                          62162  ...                                          Unerupted    
2                          62163  ...                              Sound permanent tooth    
3                          62164  ...  Permanent tooth with a dental carious surface ...    
4                          62165  ...                              Sound permanent tooth    
...                          ...  ...                                                ...    
8361                      102952  ...                              Sound permanent tooth    
8362                      102953  ...  Permanent tooth with a dental carious surface ...    
8363                      102954  ...                              Sound permanent tooth    
8364                      102955  ...  Permanent tooth with a dental carious surface ...    
8365                      102956  ...                      Missing due to dental disease    

[35909 rows x 62 columns]

In [ ]:
#c. 
print(data_clean.shape) #result from part a
print(converted_frame.shape) #result from part b

(39156, 11)
(35909, 62)


In part a, the result dataframe has 39156 rows and 11 columns. 
In part b, the result dataframe has 35909 rows and 62 columns. 